pkg/crds/crds.go

```diff
@@ -0,0 +1,205 @@
+// Package crds is used for installing rancher CRDs.
+package crds
+
+import (
+	"context"
+	"embed"
+	"fmt"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/rancher/rancher/pkg/fleet"
+	"github.com/rancher/wrangler/pkg/crd"
+	"github.com/rancher/wrangler/pkg/yaml"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	// yamlDir directory holding CRDs yaml files.
+	yamlDir = "yamls"
+
+	// readyDuration time to wait for CRDs to be ready.
+	readyDuration = time.Minute * 1
+
+	k8sManagedByKey = "app.kubernetes.io/managed-by"
+	managerValue    = "rancher"
+)
+
+//go:embed yamls/*
+var crdFS embed.FS
+
+// EnsureRequired will ensure all required CRDs needed by Rancher based on the currently enabled features are installed and up to date.
+func EnsureRequired(ctx context.Context, cfg *rest.Config) error {
+	return Ensure(ctx, cfg, RequiredCRDs())
+}
+
+// Ensure will ensure all given CRDs are installed and up to date.
+func Ensure(ctx context.Context, cfg *rest.Config, crdNames []string) error {
+	toCreateCRDs, err := getCRDs(crdNames)
+	if err != nil {
+		return fmt.Errorf("failed to get CRDs: %w", err)
+	}
+
+	ownedByRancher, err := labels.NewRequirement(k8sManagedByKey, selection.Equals, []string{managerValue})
+	if err != nil {
+		return fmt.Errorf("failed to create crd label selector: %w", err)
+	}
+	selector := labels.NewSelector().Add(*ownedByRancher)
+
+	clientSet, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create new clientset: %w", err)
+	}
+
+	err = crd.BatchCreateCRDs(ctx, clientSet.ApiextensionsV1().CustomResourceDefinitions(), selector, readyDuration, toCreateCRDs)
```

Comment:
A quick look says that none of our CRDs today are created with the "ownedByRancher" label. Does this strategy work with existing/upgraded setups (since wrangler won't see that it has existing CRDs since none match the selector)?